File: 24.两两交换链表中的节点/java/_24.java


/*import java.util.*;
import org.junit.Assert;*/
public class _Model {

    static class _1st {
        public static void main(String[] args) {
            _1st handler = new _1st();


        }

        /*
         *
         *
            dummy->   1  ->  2   ->  3   ->  4
              prev  first second    nxt
            step1:second.next = first;
            step2:first.next = nxt;
            step3:prev.next = second;
         */
        public ListNode swapPairs(ListNode head) {
            if (head == null || head.next == null) return head;
            ListNode dummy = new ListNode(-1);
            dummy.next = head;
            ListNode prev = dummy;
            while (prev.next != null && prev.next.next != null) {
                ListNode first = prev.next;
                ListNode second = prev.next.next;
                ListNode nxt = second.next;
                //step1
                second.next = first;
                //step2
                first.next = nxt;
                //step3
                prev.next = second;
                prev = first;
            }
            return dummy.next;
        }

    }


}

File: 24.两两交换链表中的节点/golang/24.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func swapPairs(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	left := head
	right := head.Next
	// 新的头节点一定从第二个位置开始
	head = right
	var pre *ListNode
	for {
		// 左节点指向right下一个位置
		left.Next = right.Next
		// 右节点指向左节点
		right.Next = left
		// pre从第一轮之后才有值
		if pre != nil {
			pre.Next = right
		}
		// pre等于本次右边的节点
		pre = left
		left = left.Next
		if left == nil || left.Next == nil {
			return head
		}
		right = left.Next
	}
	return head
}
